### ISSUE TITLE
Group2GroupCache recalculation is very slow with many groups

### ISSUE THREAD

**Issue by LucaGiamminonni** [DSpace/DSpace#8197]
The recalculation is of the Group2GroupCache done in the GroupServiceImpl.rethinkGroupCache method is very slow if there are many groups in the database (about twenty seconds with 2 thousand groups) and this slows down all the operations of update on a group when a member group has been added or removed. The situation becomes even less efficient if a process inserts or updates many groups, because after each update the cache is completely recalculated (very often unnecessarily between one update and another).

We have two proposals to solve the problem:

- do not use the cache to make direct queries on the group2group table. This would involve a greater number of queries because a query would have to be made for each level of the hierarchy, but it would increase the performance of the operations of adding and removing members. There are rare cases of very large hierarchies (I think there is usually at most one group with many subgroups) and therefore I believe that overall the performances would not be much worse. For this strategy we will need two methods to retrieve all subgroups of a given group and another to retrieve all parent groups of a group. The search can be done by making a single query for each level in the hierarchy using the IN clause with the various uuids calculated with the previous query.

- improve cache recalculation to recalculate only the portion related to the modified groups. We can also add mechanisms that allow not to force the recalculation of the cache after each update of a group so that it can only be done once at the end. This would improve the performance of processes that modify the hierarchy of many groups.

If the performances of the first solution are not low, I would opt for that solution for clarity and simplicity, rather than further complicating the existing cache. Furthermore, the existing cache also has the problem that if the developer do not force the update on a group after adding or removing a member, the cache is not recalculated.

**Comment by mwoodiupui** [comment]
I have sometimes wondered why that cache even exists.  If performance is good without it, and recalculation is slow, then I agree that we are better off without it.

Perhaps the search can be done in a single nested query or a stored procedure.  The more that can be done in the DBMS, the better.

**Comment by tdonohue** [comment]
I agree with the idea to *remove* this cache and see if our performance is good enough without it.  I think this Group2GroupCache was built at a time when there were definite performance issues.... but, we should reanalyze now whether performance is better in our current environment *without* this Group2GroupCache.  At the very least, it should give us more information on whether a custom caching solution would still be necessary in this area of the code.

I'm pulling this over to the 7.3 board for tracking purposes. If anyone is interested let me know... I would want some sort of rough estimate though if we plan to fix this via funded development

**Comment by autavares-dev** [comment]
We needed to improve the performance of group operations at my institution, due to some important changes that will be made to our com/col structure. Adding or removing a single group or subgroup was taking about 1min 30secs (final cache table with 13k rows). After the changes, it only takes 1s to 3s.

I have the related code in a personal branch (DSpace fork) [in this commit](https://github.com/autavares-dev/DSpace/commit/b6d662b728490718dfc72d99931fc8d8e5ec4e8a).

In simple terms, the new code creates a `Set<UUID, UUID>` from the current table and also for the new cache. Then, the set differences are used to delete or create individual rows in the table. In the old code, most of the CPU time is spent getting `Group` entities from UUIDs and checking if the row already exists in the cache before inserting (which seems unnecessary even in the old code).

Since we chose the second approach suggested by @LucaGiamminonni (recalculating only the part related to the modified groups), and the discussion in this issue is favoring the removal of the cache, I would like to know if this solution is still of interest for a PR.

**Comment by tdonohue** [comment]
@autavares-dev : Please do create a PR since you've found a significant way to improve the current code (and no one else has volunteered on this issue).   A PR is also useful for us to see if the code changes pass all the automated tests (if not, that's a sign there may be a bug, either in the code or in an automated test).
